src/content-routing.js

```diff
@@ -36,32 +43,76 @@ class ContentRouting {
 
   /**
    * Iterates over all content routers in series to find providers of the given key.
-   * Once a content router succeeds, iteration will stop.
    *
    * @param {CID} key - The CID key of the content to find
    * @param {object} [options]
    * @param {number} [options.timeout] - How long the query should run
    * @param {number} [options.maxNumProviders] - maximum number of providers to find
    * @returns {AsyncIterable<{ id: PeerId, multiaddrs: Multiaddr[] }>}
    */
-  async * findProviders (key, options) {
+  async * findProviders (key, options = {}) {
     if (!this.routers.length) {
       throw errCode(new Error('No content this.routers available'), 'NO_ROUTERS_AVAILABLE')
     }
 
-    const result = await pAny(
-      this.routers.map(async (router) => {
-        const provs = await all(router.findProviders(key, options))
+    /** @type Set<string> */
+    const seen = new Set()
+    const {
+      peerStore
+    } = this.libp2p
 
-        if (!provs || !provs.length) {
-          throw errCode(new Error('not found'), 'NOT_FOUND')
+    const ignoreError = async function * (source) {
+      try {
+        for await (const prov of source) {
+          yield prov
         }
-        return provs
-      })
+      } catch (err) {
+        // we will throw an error at the end if no providers have been found
+        // so log this error but otherwise ignore it
+        log.error(err)
+      }
+    }
+
+    yield * pipe(
+      merge(
+        ...this.routers.map(router => ignoreError(router.findProviders(key, options)))
+      ),
+      function filterProviders (source) {
+        return filter(source, (provider) => {
+          // ensure we have the addresses for a given peer
+          peerStore.addressBook.add(provider.id, provider.multiaddrs)
+
+          // dedupe by peer id
+          if (seen.has(provider.id.toString())) {
+            return false
+          }
+
+          seen.add(provider.id.toString())
+
+          return true
+        })
+      },
+      function returnAllAddresses (source) {
+        return map(source, (provider) => {
+          const addresses = peerStore.addressBook.get(provider.id) || []
```

Comment:
This is interesting. We might return addresses other than the ones from the router, in case we already knew the peer. Do we have any advantage with it though? For further dial, it will be good as we will attempt to dial other multiaddrs, but here it might introduce strange behaviors where the results might differ from the ones obtained with `go-ipfs`, when using the DHT.

Perhaps, we should just get the addresses from the provider record, which also means we don't need to hit the peerStore. What do you think?

Reply:
I agree, I've actually pulled this out in a further iteration.  The reason being if we've never seen the peer before (likely) and we get a second result for the same peer with different addresses (maybe likely), we'll never yield that second result because we've already yielded the first and we filter based on peer id.

Part of me thinks we shouldn't return multiaddrs at all, just the peer id, since the user now uses a peer id to connect to a remote peer instead of multiaddrs.  One for the future perhaps.

Reply:
> Part of me thinks we shouldn't return multiaddrs at all, just the peer id, since the user now uses a peer id to connect to a remote peer instead of multiaddrs. One for the future perhaps.

This seems a good call. We will be moving from returning `PeerId` to return its string identifier in the future for performance reasons. We can look into doing that as part of that breaking change in https://github.com/libp2p/js-libp2p/issues/680